### PR TITLE
Emit event on link share action

### PR DESCRIPTION
**Pull request by skjnldsv**
See https://github.com/skjnldsv/files_downloadcounter/

- [x] Sharing event
- [x] External app

Follow-up: 
- [x] Link share menu vue API
- [x] New link share menu entry to set the counter

### CODE REVIEW

File: apps/files_sharing/lib/Event/ShareLinkAccessedEvent.php [hunk 1]
```diff
@@ -0,0 +1,51 @@
+<?php
+
+declare(strict_types=1);
+/**
+ * @copyright Copyright (c) 2020 Arthur Schiwon <blizzz@arthur-schiwon.de>
```
Comment (kesselb) [inline apps/files_sharing/lib/Event/ShareLinkAccessedEvent.php:5]: Don't forget to update the year ;)

File: apps/files_sharing/lib/Event/ShareLinkAccessedEvent.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+use OCP\Share\IShare;
+
+class ShareLinkAccessedEvent extends Event {
+	/** @var IShare */
+	private $share;
+
+	/** @var string */
+	private $step;
+
+	public function __construct(IShare $share, $step = '') {
```
Comment (kesselb) [inline apps/files_sharing/lib/Event/ShareLinkAccessedEvent.php:38]:
```suggestion
	public function __construct(IShare $share, string $step = '') {
```

**Review by kesselb — APPROVED** [review]

File: apps/files_sharing/lib/Controller/ShareController.php [hunk 3]
```diff
@@ -260,19 +269,33 @@ protected function emitAccessShareHook($share, $errorCode = 200, $errorMessage =
...
+
 		if (!is_null($exception)) {
 			throw $exception;
 		}
 	}
 
+	/**
+	 * Emit a ShareLinkAccessedEvent event when a share is accessed, downloaded, auth...
+	 */
+	protected function emitShareAccessEvent(IShare $share, string $step = '', int $errorCode = 200, string $errorMessage = '') {
```
Comment (blizzz) [inline apps/files_sharing/lib/Controller/ShareController.php:290]:
```suggestion
	protected function emitShareAccessEvent(IShare $share, string $step = '', int $errorCode = 200, string $errorMessage = ''): void {
```

Comment (skjnldsv) [reply apps/files_sharing/lib/Controller/ShareController.php:290]: Now you're being picky :grin:

**Review by blizzz — APPROVED** [review]
one return type hint, otherwise :+1:

File: apps/files_sharing/lib/Controller/ShareController.php (hunk 3 above)
Comment (blizzz) [reply apps/files_sharing/lib/Controller/ShareController.php:290]: trained by cs:check :stuck_out_tongue:

**Comment by AndyScherzinger** [comment]
@skjnldsv backport to 22?

**Pull request merged**